IdentityCore/src/cache/mac/MSIDMacCredentialStorageItem.m

```diff
@@ -47,71 +47,122 @@ - (instancetype)init
     return self;
 }
 
-- (void)storeCredential:(MSIDCredentialCacheItem *)credential forKey:(MSIDCacheKey *)key
+- (void)storeItem:(id<MSIDJsonSerializable>)item inBucket:(NSString *)bucket forKey:(MSIDCacheKey *)key
 {
     dispatch_barrier_async(self.queue, ^{
-        [self.cacheObjects setObject:credential forKey:key];
+        
+        NSMutableDictionary *items = nil;
+        if (bucket && ![self.cacheObjects objectForKey:bucket])
+        {
+            items = [NSMutableDictionary new];
+        }
+        else
+        {
+            items = [self.cacheObjects objectForKey:bucket];
+        }
+        
+        [items setObject:item forKey:key];
+        [self.cacheObjects setObject:items forKey:bucket];
     });
 }
 
-- (void)mergeStorageItem:(MSIDMacCredentialStorageItem *)storageItem
+- (void)mergeStorageItem:(MSIDMacCredentialStorageItem *)storageItem inBucket:(NSString *)bucket
 {
     dispatch_barrier_async(self.queue, ^{
-        for (MSIDCacheKey *key in storageItem.cacheObjects)
+        
+        for (NSString *bucketKey in storageItem.cacheObjects)
         {
-            MSIDCredentialCacheItem *credential = [storageItem.cacheObjects objectForKey:key];
-            if (credential)
+            NSMutableDictionary *bucketDict = [storageItem.cacheObjects objectForKey:bucketKey];
+            NSMutableDictionary *subDict = [self.cacheObjects objectForKey:bucketKey];
+            
+            if (!subDict)
             {
-                [self.cacheObjects setObject:credential forKey:key];
+                [self.cacheObjects setObject:bucketDict forKey:bucketKey];
             }
             else
             {
-                MSID_LOG_WITH_CTX_PII(MSIDLogLevelError, nil, @"Credential is nil for key %@ while merging storage credentials.", MSID_PII_LOG_MASKABLE(key));
+                for (MSIDCacheKey *key in bucketDict)
+                {
+                    id<MSIDJsonSerializable> storedItem = [bucketDict objectForKey:key];
+                    if (storedItem)
+                    {
+                        [subDict setObject:storedItem forKey:key];
+                    }
+                    else
+                    {
+                        MSID_LOG_WITH_CTX_PII(MSIDLogLevelError, nil, @"Item is nil for key %@ while merging storage item.", MSID_PII_LOG_MASKABLE(key));
+                    }
+                }
+                
+                [self.cacheObjects setObject:subDict forKey:bucketKey];
             }
         }
     });
 }
 
-- (void)removeStoredCredentialForKey:(MSIDCacheKey *)key
+- (void)removeStoredItemForKey:(MSIDCacheKey *)key inBucket:(NSString *)bucket
 {
     dispatch_barrier_sync(self.queue, ^{
-        [self.cacheObjects removeObjectForKey:key];
+        if (bucket && [self.cacheObjects objectForKey:bucket])
```

Comment:
nit: else log

Reply:
I think we can avoid logging here. If we are trying to remove an item and it does not exist, should we still log? Please let me know if you think we still need the log statement.